crypto/bio/bio_lib.c

```diff
@@ -786,41 +786,48 @@ void bio_cleanup(void)
 }
 
 /* Internal variant of the below BIO_wait() not calling BIOerr() */
-static int bio_wait(BIO *bio, time_t max_time, unsigned int milliseconds)
+static int bio_wait(BIO *bio, time_t max_time, unsigned int fallback_millisec)
 {
 #ifndef OPENSSL_NO_SOCK
     int fd;
 #endif
+    long sec_diff;
 
     if (max_time == 0)
         return 1;
 
 #ifndef OPENSSL_NO_SOCK
-    if (BIO_get_fd(bio, &fd) > 0)
+    if (BIO_get_fd(bio, &fd) > 0 && fd < FD_SETSIZE)
         return BIO_socket_wait(fd, BIO_should_read(bio), max_time);
 #endif
-    if (milliseconds > 1000) {
-        long sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    /* fall back to polling since no sockets are available */
 
-        if (sec_diff <= 0)
-            return 0; /* timeout */
-        if ((unsigned long)sec_diff < milliseconds / 1000)
-            milliseconds = (unsigned long)sec_diff * 1000;
+    sec_diff = (long)(max_time - time(NULL)); /* might overflow */
+    if (sec_diff < 0)
+        return 0; /* clearly timeout */
+
+    /* now wait at most the given number of milliseconds */
+    if (fallback_millisec > 1000) {
+
+        if (sec_diff < fallback_millisec / 1000)
+            fallback_millisec = sec_diff * 1000;
```

Comment:
I did not mean no typecast at all.
I just wanted to say that I would be more appropriate to do
```C
if (sec_diff < fallback_millisec / 1000)
    fallback_millisec = (unsigned int) sec_diff * 1000;
```

Reply:
whee, so make that
```C
if ((unsigned long)sec_diff < fallback_millisec / 1000)
    fallback_millisec = (unsigned int) sec_diff * 1000;
```

Reply:
Done